### PR TITLE
Simplify C# random sort

### PR DESCRIPTION
**Pull request by pepone**
This tiny PR fixes missing random sorting of AdapterID in C# IceDiscovery implementation, it also simplifies the random sorting of endpoints in Reference.cs

### CODE REVIEW

**Comment by pepone** [comment]
Maybe we should add an extension method that does random sorting and use in both places, what do you think?

**Comment by bernardnormier** [comment]
Locking _rand for each iteration seems inefficient. Could you lock it for the entire iteration? And then, where - around to the ToArray call?
Should we combine the two OrderBy in Reference?

**Comment by pepone** [comment]
> Locking _rand for each iteration seems inefficient. Could you lock it for the entire iteration? And then, where - around to the ToArray call?
> Should we combine the two OrderBy in Reference?

Move this to Shuffle extension method and fixed the lock

**Comment by bernardnormier** [comment]
There is a still a lock per iteration - what did you fix?

I don't see how moving this code to an IEnumerable extension could allow us to reduce the number of lock acquire/release.

**Comment by bernardnormier** [comment]
The nicest solution could be to keep the extension and but make `_rand` thread-static, as outlined in https://devblogs.microsoft.com/pfxteam/getting-random-numbers-in-a-thread-safe-way/.

**Comment by bernardnormier** [comment]
> There is a still a lock per iteration - what did you fix?

I misread the code. There is now a lock per-call (good), but I am concerned OrderBy is lazily evaluated and as a result `_rand.Next()` gets called later with `_rand` not locked.

File: csharp/src/Ice/EnumerableExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    public static class EnumerableExtensions
+    {
+        private static readonly ThreadLocal<Random> _rand = new ThreadLocal<Random>(() => new Random());
+
+        /// <summary>Shuffle the elements of a sequence.</summary>
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <param name="source">A sequence of values to shuffle.</param>
+        /// <returns>An IEnumerable&lt;TSouce&gt; whose elements are shuffled.</returns>
+        public static IEnumerable<TSource> Shuffle<TSource>(this IEnumerable<TSource> source)
+        {
```
Comment (bernardnormier) [inline csharp/src/Ice/EnumerableExtensions.cs:22]: Could we simplify this to `=> source.OrderBy(element => _rand.Value!.Next());`?

File: csharp/src/Ice/EnumerableExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+using System.Linq;
+using System.Diagnostics;
+
+namespace Ice
+{
+    public static class EnumerableExtensions
+    {
+        private static readonly ThreadLocal<Random> _rand = new ThreadLocal<Random>(() => new Random());
+
+        /// <summary>Shuffle the elements of a sequence.</summary>
```
Comment (bernardnormier) [inline csharp/src/Ice/EnumerableExtensions.cs:17]: Should be Shuffles.

File: csharp/src/Ice/EnumerableExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <param name="source">A sequence of values to shuffle.</param>
+        /// <returns>An IEnumerable&lt;TSouce&gt; whose elements are shuffled.</returns>
+        public static IEnumerable<TSource> Shuffle<TSource>(this IEnumerable<TSource> source)
```
Comment (bernardnormier) [inline csharp/src/Ice/EnumerableExtensions.cs:21]: Use a cref attribute with {} to reference the other type. See https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/cref-attribute

**Pull request merged**